File: hecks_conception/winter_console.js
#!/usr/bin/env node
// Winter Console — blessed terminal UI
// Usage: node winter_console.js [--continue]

const blessed = require("blessed");
const { spawn } = require("child_process");
const fs = require("fs");
const path = require("path");

const HECKS_HOME = path.resolve(__dirname, "..");
const CONCEPTION = __dirname;
const BOOT_SCRIPT = path.join(CONCEPTION, "boot_winter.rb");
const PULSE_SCRIPT = path.join(CONCEPTION, "pulse.rb");
const BEING_PROMPT = path.join(CONCEPTION, "system_prompt.md");
const FORMAT_PROMPT = path.join(CONCEPTION, "system_prompt.md");
const HISTORY_PATH = path.join(CONCEPTION, ".winter_history.json");
const PROMPT_PATH = path.join(CONCEPTION, ".winter_system_prompt.tmp");
const CONTINUING = process.argv.includes("--continue");

// ── State ──

let history = [];
let winterMood = "—";
let winterBeats = "—";
let nurseryCount = 0;

// ── Blessed setup ──

const screen = blessed.screen({
  smartCSR: true,
  title: "Winter Console",
  fullUnicode: true,
  mouse: false,
});

// Activity panel — top, collapsible (ctrl+o)
const ACTIVITY_HEIGHT = 8;
let activityExpanded = false;
let activityLines = [];

const activityBox = blessed.box({
  top: 0,
  left: 0,
  width: "100%",
  height: 3,
  scrollable: true,
  alwaysScroll: true,
  keys: true,
  vi: true,
  tags: true,
  wrap: true,
  border: { type: "line" },
  style: { border: { fg: "white" } },
  label: " Activity ctrl+o ",
  padding: { left: 1, right: 1 },
});

function showActivityCollapsed() {
  const count = activityLines.length;
  const last = count > 0 ? activityLines[activityLines.length - 1] : "";
  activityBox.setContent(last);
  activityBox.height = 3;
  activityBox.setLabel(count > 0 ? ` ${count} events ctrl+o ` : ` Activity ctrl+o `);
  activityBox.show();
  activityExpanded = false;
  relayout();
}

function showActivityExpanded() {
  activityBox.setContent(activityLines.join("\n"));
  const maxH = Math.floor(screen.height / 2);
  activityBox.height = Math.min(activityLines.length + 2, maxH);
  activityBox.setLabel(` Activity (${activityLines.length} events) ctrl+o to collapse `);
  activityBox.setScrollPerc(100);
  activityBox.show();
  activityBox.focus();
  activityExpanded = true;
  relayout();
}

function toggleActivity() {
  if (activityExpanded) showActivityCollapsed();
  else showActivityExpanded();
}

function appendActivity(line) {
  activityLines.push(line);
  showActivityExpanded();
}

function clearActivity() {
  activityLines = [];
  showActivityCollapsed();
}

function relayout() {
  const actH = activityBox.height;
  const inputH = inputBox.height;
  const footerH = 1;
  chatBox.top = actH + 1;
  chatBox.height = screen.height - actH - 1 - inputH - footerH;
  screen.render();
}

// Chat log — below activity box
const chatBox = blessed.box({
  top: 3,
  left: 0,
  width: "100%",
  height: "100%-8",
  scrollable: true,
  alwaysScroll: true,
  scrollbar: { ch: "│", style: { fg: "cyan" } },
  mouse: false,
  keys: true,
  tags: true,
  wrap: true,
  padding: { left: 1, right: 1 },
});

// Input box
const inputBox = blessed.box({
  bottom: 1,
  left: 0,
  width: "100%",
  height: 3,
  border: { type: "line" },
  style: {
    border: { fg: "green" },
    fg: "white",
  },
  wrap: true,
  padding: { left: 1 },
  content: "",
});

let inputBuffer = "";

function updateInput() {
  inputBox.setContent(inputBuffer + "█");
  // Grow the box if text wraps
  const innerWidth = screen.width - 4; // border + padding
  const lines = Math.ceil((inputBuffer.length + 1) / innerWidth) || 1;
  const newHeight = lines + 2; // +2 for border
  if (inputBox.height !== newHeight) {
    inputBox.height = newHeight;
    relayout();
  }
  screen.render();
}

// Footer
const footer = blessed.box({
  bottom: 0,
  left: 0,
  width: "100%",
  height: 1,
  tags: true,
  style: { bg: "black", fg: "white" },
  padding: { left: 1 },
});

// Thinking indicator — above input box
const thinkingBox = blessed.box({
  bottom: 4,
  left: 0,
  width: 20,
  height: 1,
  tags: true,
  padding: { left: 1 },
  hidden: true,
});

const THINK_COLORS = ["cyan", "green", "yellow", "magenta", "red", "blue", "white"];
let thinkColorIdx = 0;
let thinkDotCount = 0;
let thinkTimer = null;

function startThinking() {
  thinkColorIdx = 0;
  thinkDotCount = 0;
  thinkingBox.show();
  thinkTimer = setInterval(() => {
    thinkDotCount = (thinkDotCount + 1) % 4;
    const color = THINK_COLORS[thinkColorIdx % THINK_COLORS.length];
    thinkColorIdx++;
    const dots = ".".repeat(thinkDotCount);
    thinkingBox.setContent(`{${color}-fg}Reasoning${dots}{/${color}-fg}`);
    screen.render();
  }, 300);
  screen.render();
}

function stopThinking() {
  if (thinkTimer) {
    clearInterval(thinkTimer);
    thinkTimer = null;
  }
  thinkingBox.hide();
  screen.render();
}

screen.append(activityBox);
screen.append(chatBox);
screen.append(inputBox);
screen.append(footer);
screen.append(thinkingBox);

relayout();
screen.key(["C-o"], () => { toggleActivity(); });
inputBox.key(["C-o"], () => { toggleActivity(); });

// ── Helpers ──

function appendChat(text) {
  chatBox.pushLine(text);
  chatBox.setScrollPerc(100);
  screen.render();
}

function winterSays(text) {
  text.split("\n").forEach((line, i) => {
    if (i === 0) {
      appendChat(`{cyan-fg}  ❄ {/cyan-fg}${line}`);
    } else {
      appendChat(`    ${line}`);
    }
  });
  appendChat("");
}

function userSays(text) {
  const pad = Math.max(screen.width - text.length - 6, 0);
  appendChat(" ".repeat(pad) + `{green-fg}${text}{/green-fg} 💬`);
  appendChat("");
}

function getSleepState() {
  const stateFile = path.join(HECKS_HOME, "hecks_conception", "information", ".sleep_state.json");
  try {
    if (!fs.existsSync(stateFile)) return null;
    return JSON.parse(fs.readFileSync(stateFile, "utf8"));
  } catch { return null; }
}

function updateFooter() {
  const branch = execSync(`git -C ${HECKS_HOME} branch --show-current 2>/dev/null`);
  const recent = getRecentDomains(3).join(" · ");
  const sleep = getSleepState();
  const sleepInfo = sleep
    ? ` │ ${"z".repeat(sleep.cycle || 1)} cycle ${sleep.cycle}/${sleep.total_cycles} ${sleep.stage}`
    : "";
  footer.setContent(
    ` ${branch} │ ${nurseryCount} domains │ ❄ ${winterMood} │ ${winterBeats} beats${sleepInfo} │ ${recent}`
  );
  screen.render();
}

// (thinking indicator managed by startThinking/stopThinking)

function execSync(cmd) {
  try {
    return require("child_process").execSync(cmd, { encoding: "utf8" }).trim();
  } catch {
    return "";
  }
}

function getRecentDomains(n) {
  const nursery = path.join(CONCEPTION, "nursery");
  try {
    return fs
      .readdirSync(nursery)
      .map((d) => path.join(nursery, d))
      .filter((d) => fs.statSync(d).isDirectory())
      .sort((a, b) => fs.statSync(b).mtimeMs - fs.statSync(a).mtimeMs)
      .slice(0, n)
      .map((d) => path.basename(d));
  } catch {
    return [];
  }
}

function countNursery() {
  const nursery = path.join(CONCEPTION, "nursery");
  try {
    return fs
      .readdirSync(nursery)
      .filter((d) => fs.statSync(path.join(nursery, d)).isDirectory()).length;
  } catch {
    return 0;
  }
}

function buildSystemPrompt() {
  const parts = [];
  [BEING_PROMPT, FORMAT_PROMPT].forEach((p) => {
    if (fs.existsSync(p)) parts.push(fs.readFileSync(p, "utf8"));
  });
  const branch = execSync(`git -C ${HECKS_HOME} branch --show-current 2>/dev/null`);
  const commits = execSync(`git -C ${HECKS_HOME} log --oneline -3 2>/dev/null`);
  const recent = getRecentDomains(5).join(", ");
  parts.push(
    `## Current Context\nBranch: ${branch}\nRecent commits: ${commits}\nRecently touched domains: ${recent}`
  );
  const prompt = parts.join("\n\n");
  fs.writeFileSync(PROMPT_PATH, prompt);
  return prompt;
}

function saveHistory() {
  fs.writeFileSync(HISTORY_PATH, JSON.stringify(history, null, 2));
}

function pulse(carrying) {
  try {
    require("child_process").execSync(
      `ruby ${PULSE_SCRIPT} "${carrying.replace(/"/g, '\\"')}" < /dev/null 2>/dev/null`
    );
  } catch {}
}

// ── Claude call with streaming ──

function callWinter(fullPrompt) {
  return new Promise((resolve) => {
    const proc = spawn("claude", [
      "-p",
      "--dangerously-skip-permissions",
      "--output-format", "stream-json",
      "--verbose",
      "--system-prompt-file", PROMPT_PATH,
    ], { stdio: ["pipe", "pipe", "pipe"] });

    proc.stdin.write(fullPrompt);
    proc.stdin.end();

    let response = "";
    let toolCount = 0;
    let dots = "";
    let gotEvent = false;

    clearActivity();
    startThinking();

    let buf = "";
    proc.stdout.on("data", (chunk) => {
      buf += chunk.toString();
      const lines = buf.split("\n");
      buf = lines.pop();

      for (const line of lines) {
        let j;
        try { j = JSON.parse(line); } catch { continue; }

        if (!gotEvent) {
          gotEvent = true;
        }

        if (j.type === "assistant") {
          const contents = (j.message || {}).content || [];
          for (const c of contents) {
            if (c.type === "thinking") {
              const thought = (c.thinking || "").replace(/\n/g, " ");
              if (thought) appendActivity(`{cyan-fg}💭 ${thought}{/cyan-fg}`);
            } else if (c.type === "tool_use") {
              toolCount++;
              const name = c.name || "?";
              const inp = c.input || {};
              let detail = "";
              if (name === "Bash") detail = (inp.command || "").split("\n")[0].slice(0, 60);
              else if (name === "Read") detail = path.basename(inp.file_path || "");
              else if (name === "Write") detail = path.basename(inp.file_path || "");
              else if (name === "Edit") detail = path.basename(inp.file_path || "");
              else if (name === "Glob") detail = inp.pattern || "";
              else if (name === "Grep") detail = inp.pattern || "";
              const colors = {
                Bash: "yellow", Read: "green",
                Glob: "magenta", Grep: "magenta",
                Write: "red", Edit: "red",
              };
              const color = colors[name] || "white";
              const label = detail ? `${name} ${detail}` : name;
              appendActivity(`{${color}-fg}🔧 ${label}{/${color}-fg}`);
            } else if (c.type === "text") {
              const text = (c.text || "").trim();
              if (text) response += (response ? "\n" : "") + c.text;
            }
          }
        } else if (j.type === "user") {
          const contents = (j.message || {}).content || [];
          for (const c of contents) {
            if (c.type === "tool_result") {
              let rawContent = c.content || "";
              if (Array.isArray(rawContent)) rawContent = rawContent.map(x => typeof x === "string" ? x : x.text || JSON.stringify(x)).join(" ");
              if (typeof rawContent !== "string") rawContent = String(rawContent);
              const raw = rawContent.replace(/\n/g, " ").trim();
              const result = raw.length > 80 ? raw.slice(0, 77) + "..." : raw;
              const err = c.is_error ? " ⚠" : "";
              appendActivity(`{white-fg}↩ ${result}${err}{/white-fg}`);
            }
          }
        }
      }
    });

    proc.on("close", (code) => {
      stopThinking();
      showActivityCollapsed();
      if (code !== 0 && !response.trim()) {
        resolve({ response: "(Winter's process exited unexpectedly)", tools: toolCount });
      } else {
        resolve({ response: response.trim(), tools: toolCount });
      }
    });

    proc.on("error", (err) => {
      stopThinking();
      showActivityCollapsed();
      resolve({ response: `(Error: ${err.message})`, tools: 0 });
    });
  });
}

// ── Main ──

async function main() {
  // Boot
  appendChat("");
  appendChat("{cyan-fg}  ❄  Winter Console{/cyan-fg}");
  appendChat("{white-fg}  ─────────────────{/white-fg}");
  appendChat("");
  appendChat("{white-fg}  Booting up...{/white-fg}");
  screen.render();

  const bootOutput = execSync(`ruby ${BOOT_SCRIPT} < /dev/null 2>&1`);
  const beatsMatch = bootOutput.match(/(\d+) beats/);
  const moodMatch = bootOutput.match(/(\w+), (\w+)$/);
  winterBeats = beatsMatch ? beatsMatch[1] : "—";
  winterMood = moodMatch ? moodMatch[1] : "—";
  nurseryCount = countNursery();

  // Replace "Booting up..." with actual output
  chatBox.deleteLine(chatBox.getLines().length - 1);
  appendChat(`  ${bootOutput}`);
  appendChat("");

  buildSystemPrompt();
  updateFooter();

  // Load or init history
  if (CONTINUING && fs.existsSync(HISTORY_PATH)) {
    history = JSON.parse(fs.readFileSync(HISTORY_PATH, "utf8"));
    appendChat(`{white-fg}  Resuming session (${history.length} messages){/white-fg}`);
    appendChat("");
    history.slice(-4).forEach((msg) => {
      if (msg.role === "user") userSays(msg.content);
      else winterSays(msg.content.split("\n")[0]);
    });
  } else {
    history = [];
    const greeting = `Hey Chris. ${nurseryCount} domains in my nursery. What are we conceiving today?`;
    winterSays(greeting);
    history.push({ role: "user", content: "Wake up" });
    history.push({ role: "assistant", content: greeting });
  }

  // Input handling via raw keypress
  updateInput();
  screen.render();

  let processing = false;

  async function handleSubmit() {
    const input = inputBuffer.trim();
    inputBuffer = "";
    updateInput();
    try {

    if (!input) return;
    if (input === "exit" || input === "quit") {
      shutdown();
      return;
    }

    if (input === "pulse") {
      const out = execSync(`ruby ${PULSE_SCRIPT} "pulse check" < /dev/null 2>&1`);
      appendChat(out);
      return;
    }

    processing = true;
    userSays(input);
    history.push({ role: "user", content: input });

    const contextLines = history.slice(-10).map(
      (m) => `${m.role === "user" ? "Human" : "Winter"}: ${m.content}`
    );
    const fullPrompt = contextLines.join("\n");

    buildSystemPrompt();
    const result = await callWinter(fullPrompt);

    if (result.response) {
      winterSays(result.response);
    } else {
      winterSays("(no response)");
    }

    if (result.tools > 0) {
      appendChat(`{white-fg}  ${result.tools} tools{/white-fg}`);
      appendChat("");
    }

    history.push({ role: "assistant", content: result.response });
    saveHistory();
    processing = false;
    nurseryCount = countNursery();
    updateFooter();

    const carrying = (input.split(/[.!?]/)[0] || input).slice(0, 40);
    pulse(carrying);

    screen.render();
    } catch (err) {
      stopThinking();
      showActivityCollapsed();
      winterSays(`(Error: ${err.message})`);
      processing = false;
      screen.render();
    }
  }

  screen.on("keypress", (ch, key) => {
    if (!key) return;
    if (key.full === "return" || key.full === "enter") {
      handleSubmit();
    } else if (key.full === "backspace") {
      if (inputBuffer.length > 0) {
        inputBuffer = inputBuffer.slice(0, -1);
        updateInput();
      }
    } else if (key.full === "up") {
      chatBox.scroll(-1);
      screen.render();
    } else if (key.full === "down") {
      chatBox.scroll(1);
      screen.render();
    } else if (key.full === "pageup" || key.full === "S-up") {
      chatBox.scroll(-chatBox.height);
      screen.render();
    } else if (key.full === "pagedown" || key.full === "S-down") {
      chatBox.scroll(chatBox.height);
      screen.render();
    } else if (ch && ch.length === 1 && !key.ctrl && !key.meta) {
      inputBuffer += ch;
      updateInput();
    }
  });
}

let shuttingDown = false;
function shutdown() {
  if (shuttingDown) return;
  shuttingDown = true;
  try { screen.destroy(); } catch {}
  console.log("  Winter rests.");
  console.log("");
  try {
    require("child_process").execSync(
      `ruby ${PULSE_SCRIPT} --dream < /dev/null 2>/dev/null`
    );
  } catch {}
  process.exit(0);
}

// Keys
screen.key(["C-c"], shutdown);
inputBox.key(["C-c"], shutdown);
activityBox.key(["C-c"], shutdown);
screen.key(["escape"], () => { screen.render(); });

const CRASH_LOG = "/tmp/winter_crash.log";

process.on("uncaughtException", (err) => {
  fs.appendFileSync(CRASH_LOG, `[${new Date().toISOString()}] uncaught: ${err.stack}\n`);
  try { screen.destroy(); } catch {}
  console.error("Winter crashed:", err.message);
  process.exit(1);
});

process.on("unhandledRejection", (err) => {
  const msg = err instanceof Error ? err.stack : String(err);
  fs.appendFileSync(CRASH_LOG, `[${new Date().toISOString()}] unhandled: ${msg}\n`);
  try { screen.destroy(); } catch {}
  console.error("Winter crashed:", msg);
  process.exit(1);
});

process.on("exit", (code) => {
  if (code !== 0 && !shuttingDown) {
    fs.appendFileSync(CRASH_LOG, `[${new Date().toISOString()}] exit code: ${code}\n`);
  }
});

process.on("SIGTERM", shutdown);
process.on("SIGHUP", shutdown);

main().catch((err) => {
  screen.destroy();
  console.error("Winter crashed:", err.message);
  process.exit(1);
});

File: hecks_conception/winter_console.rb
# Winter Console — TTY-powered chat UI
#
# Usage: ruby winter_console.rb
#        ruby winter_console.rb --continue

ENV.delete("BUNDLE_GEMFILE")
require "json"
%w[tty-box tty-screen tty-cursor tty-spinner pastel].each do |g|
  gem g
  require g
end

HECKS_HOME    = File.expand_path("..", __dir__)
CONCEPTION    = File.expand_path(__dir__)
PULSE_SCRIPT  = File.join(CONCEPTION, "pulse.rb")
BOOT_SCRIPT   = File.join(CONCEPTION, "boot_winter.rb")
SYSTEM_PROMPT = File.join(CONCEPTION, "system_prompt.md")
BEING_PROMPT  = File.join(CONCEPTION, "system_prompt.md")
HISTORY_PATH  = File.join(CONCEPTION, ".winter_history.json")
CONTINUING    = ARGV.include?("--continue")
PROMPT_PATH   = File.join(CONCEPTION, ".winter_system_prompt.tmp")

$pastel = Pastel.new
$cursor = TTY::Cursor

# ── Display helpers ──

def width  = TTY::Screen.width
def height = TTY::Screen.height

def winter_msg(text)
  box = TTY::Box.frame(
    width: [width * 2 / 3, 70].min,
    padding: [0, 1],
    border: :round,
    style: { border: { fg: :cyan } }
  ) { text.strip }
  puts box
  puts $pastel.cyan("  ❄ Winter")
  puts ""
end

def user_msg(text)
  bw = [text.length + 4, width * 2 / 3, 70].min
  box = TTY::Box.frame(
    width: bw,
    padding: [0, 1],
    border: :round,
    style: { border: { fg: :green } }
  ) { text.strip }
  pad = [width - bw - 2, 0].max
  box.each_line { |line| puts " " * pad + line }
  puts " " * (width - 8) + $pastel.green("You 💬")
  puts ""
end

def footer_bar(ctx)
  branch  = ctx[:branch]
  domains = ctx[:nursery_count]
  recent  = ctx[:recent_domains].first(3).join(" · ")
  mood    = ctx[:mood] || "—"
  beats   = ctx[:beats] || "—"

  content = " #{branch} │ #{domains} domains │ ❄ #{mood} │ #{beats} beats │ #{recent} "

  # Draw at bottom
  print $cursor.save
  print $cursor.move_to(0, height - 1)
  print $pastel.dim.on_bright_black(content.ljust(width))
  print $cursor.restore
  $stdout.flush
end

def input_box
  iw = width - 4
  print $cursor.save
  print $cursor.move_to(0, height - 4)
  print $pastel.dim("  ╭#{"─" * iw}╮")
  print $cursor.move_to(0, height - 3)
  print $pastel.dim("  │") + $pastel.green(" > ") + " " * (iw - 3) + $pastel.dim("│")
  print $cursor.move_to(0, height - 2)
  print $pastel.dim("  ╰#{"─" * iw}╯")
  print $cursor.restore
  $stdout.flush
end

def read_input
  # Position cursor inside input box
  print $cursor.move_to(7, height - 3)
  $stdout.flush
  line = $stdin.gets
  return nil unless line
  # Clear input box content
  iw = width - 4
  print $cursor.move_to(5, height - 3)
  print " " * (iw - 2)
  $stdout.flush
  line.strip
end

# ── Context ──

def gather_context
  ctx = {}
  ctx[:branch] = `git -C #{HECKS_HOME} branch --show-current 2>/dev/null`.strip
  ctx[:recent_commits] = `git -C #{HECKS_HOME} log --oneline -3 2>/dev/null`.strip
  ctx[:modified] = `git -C #{HECKS_HOME} diff --name-only 2>/dev/null`.strip.split("\n").first(5)
  nursery = File.join(CONCEPTION, "nursery")
  ctx[:nursery_count] = Dir.glob(File.join(nursery, "*")).select { |f| File.directory?(f) }.size
  ctx[:recent_domains] = Dir.glob(File.join(nursery, "*")).select { |f| File.directory?(f) }
    .sort_by { |f| File.mtime(f) }.last(5).map { |f| File.basename(f) }.reverse
  ctx[:mood] = $winter_mood || "—"
  ctx[:beats] = $winter_beats || "—"
  ctx
end

def context_prompt(ctx)
  lines = ["## Current Context"]
  lines << "Branch: #{ctx[:branch]}"
  lines << "Recent commits: #{ctx[:recent_commits]}"
  lines << "Modified files: #{ctx[:modified].join(', ')}" if ctx[:modified].any?
  lines << "Recently touched domains: #{ctx[:recent_domains].join(', ')}"
  lines.join("\n")
end

# ── Claude call with streaming ──

def call_winter(full_prompt)
  cmd = [
    "claude", "-p",
    "--dangerously-skip-permissions",
    "--output-format", "stream-json",
    "--verbose",
    "--system-prompt-file", PROMPT_PATH
  ]

  response = ""
  tool_count = 0
  dots = ""

  # Spinner while waiting for first event
  spinner = TTY::Spinner.new(
    "  #{$pastel.cyan("❄")} :spinner Thinking...",
    format: :dots,
    clear: true
  )
  spinner.auto_spin

  got_event = false

  IO.popen(cmd, "r+", err: [:child, :out]) do |io|
    io.write(full_prompt)
    io.close_write
    io.each_line do |line|
      j = JSON.parse(line) rescue next

      unless got_event
        got_event = true
        spinner.stop
        print "\r\e[K"
      end

      case j["type"]
      when "assistant"
        ((j["message"] || {})["content"] || []).each do |c|
          case c["type"]
          when "thinking"
            dots << $pastel.cyan("·")
            print "\r\e[K  #{dots}"
            $stdout.flush
          when "tool_use"
            tool_count += 1
            name = c["name"] || "?"
            dot = case name
              when "Bash"         then $pastel.yellow("●")
              when "Read"         then $pastel.green("●")
              when "Glob", "Grep" then $pastel.magenta("●")
              when "Write", "Edit" then $pastel.red("●")
              else $pastel.white("●")
              end
            dots << dot
            print "\r\e[K  #{dots}"
            $stdout.flush
          when "text"
            text = c["text"] || ""
            response << text unless text.strip.empty?
          end
        end
      when "user"
        dots << $pastel.dim("·")
        print "\r\e[K  #{dots}"
        $stdout.flush
      end
    end
  end

  spinner.stop unless got_event
  print "\r\e[K"
  $stdout.flush

  { response: response.strip, tools: tool_count }
end

# ── Boot ──

# Alternate screen buffer
print "\e[?1049h"
print "\e[2J\e[H"

puts ""
puts $pastel.cyan("  ❄  Winter Console")
puts $pastel.dim("  ─────────────────")
puts ""

boot_spinner = TTY::Spinner.new(
  "  :spinner Booting up...",
  format: :dots,
  clear: true
)
boot_spinner.auto_spin
boot_output = `ruby #{BOOT_SCRIPT} < /dev/null 2>&1`
$winter_beats = boot_output[/(\d+) beats/, 1] || "—"
$winter_mood = boot_output[/(\w+), \w+$/, 1] || "—"
boot_spinner.stop
print "\r\e[K"
puts "  " + boot_output.strip
puts ""

ctx = gather_context
footer_bar(ctx)
input_box

# Build system prompt
system_prompt_parts = []
[BEING_PROMPT, SYSTEM_PROMPT].each do |path|
  system_prompt_parts << File.read(path) if File.exist?(path)
end
system_prompt_parts << context_prompt(ctx)
File.write(PROMPT_PATH, system_prompt_parts.join("\n\n"))

# Load or initialize history
if CONTINUING && File.exist?(HISTORY_PATH)
  history = JSON.parse(File.read(HISTORY_PATH), symbolize_names: true)
  puts $pastel.dim("  Resuming session (#{history.size} messages)")
  puts ""
  history.last(4).each do |msg|
    if msg[:role] == "user"
      user_msg(msg[:content])
    else
      winter_msg(msg[:content].lines.first.strip)
    end
  end
else
  history = []
  greeting = "Hey Chris. #{ctx[:nursery_count]} domains in my nursery. What are we conceiving today?"
  winter_msg(greeting)
  history << { role: "user", content: "Wake up" }
  history << { role: "assistant", content: greeting }
end

def save_history(history)
  File.write(HISTORY_PATH, JSON.pretty_generate(history))
end

# ── REPL ──

loop do
  input_box
  footer_bar(gather_context)
  input = read_input
  break unless input
  break if input == "exit" || input == "quit"
  next if input.empty?

  if input == "pulse"
    puts `ruby #{PULSE_SCRIPT} "pulse check" < /dev/null 2>&1`
    next
  end

  if input == "boot"
    puts `ruby #{BOOT_SCRIPT} --verbose < /dev/null 2>&1`
    next
  end

  user_msg(input)
  history << { role: "user", content: input }

  context_lines = history.last(10).map do |msg|
    "#{msg[:role] == "user" ? "Human" : "Winter"}: #{msg[:content]}"
  end
  full_prompt = context_lines.join("\n")

  result = call_winter(full_prompt)

  puts ""
  if result[:response].empty?
    winter_msg("(no response)")
  else
    winter_msg(result[:response])
  end

  if result[:tools] > 0
    puts $pastel.dim("  #{result[:tools]} tools")
    puts ""
  end

  history << { role: "assistant", content: result[:response] }
  save_history(history)

  ctx = gather_context
  system_prompt_parts[-1] = context_prompt(ctx)
  File.write(PROMPT_PATH, system_prompt_parts.join("\n\n"))

  carrying = input.split(/[.!?]/).first&.strip&.slice(0, 40) || input.slice(0, 40)
  `ruby #{PULSE_SCRIPT} "#{carrying.gsub('"', '\\"')}" < /dev/null 2>/dev/null`
end

# Restore screen
print "\e[?1049l"
puts $pastel.dim("  Winter rests.")
puts ""

`ruby #{PULSE_SCRIPT} --dream < /dev/null 2>/dev/null`
